### PR TITLE
Nginx, kafka, celery support for Docker. 

### PR DESCRIPTION
**Pull request by philipkaare**
This PR adds support for Nginx, kafka and celery. It adds two modes to dockerhq.sh, a production and development, each with a different localsettings. The proxy only runs in production mode.

### CODE REVIEW

**Comment by dimagimon** [comment]
There are 7 errors:

* docker/localsettings_docker.py, line 12 - E261 at least two spaces before inline comment
* docker/localsettings_docker.py, line 48 - E501 line too long (383 > 115 characters)
* docker/localsettings_docker.py, line 49 - W391 blank line at end of file
* docker/localsettings_docker_prod.py, line 7 - E225 missing whitespace around operator
* docker/localsettings_docker_prod.py, line 12 - E261 at least two spaces before inline comment
* docker/localsettings_docker_prod.py, line 53 - E501 line too long (383 > 115 characters)
* docker/localsettings_docker_prod.py, line 54 - W391 blank line at end of file

File: dev_settings.py [hunk 1]
```diff
@@ -75,8 +75,7 @@
 CSRF_SOFT_MODE = False
 
 # Make a dir to use for storing attachments as blobs on the filesystem
-shared_dirname = os.path.join(os.path.dirname(os.path.realpath(__file__)),
-                              'sharedfiles')
+shared_dirname = '/mnt/sharedfiles'
```
Comment (snopoke) [inline dev_settings.py]: this will screw up other people's local setups

Comment (snopoke) [inline dev_settings.py]: The `dockersettings.py` file already defines the `SHARED_DRIVE_ROOT` settings as `/sharedfiles`

File: docker/bootstrap.sh [hunk 2]
```diff
@@ -2,9 +2,27 @@
...
+
+echo "Please enter admin password"
+read ADMIN_PASSWORD
+
+echo "OK. Starting services and then bootstrapping. "
+
 $DOCKER_DIR/docker-services.sh start
 
-web_runner run --rm web /mnt/docker/bootstrap_internal.sh
+#web_runner run --rm web /mnt/docker/bootstrap_internal.sh $PROJECT_NAME $ADMIN_EMAIL $ADMIN_PASSWORD
```
Comment (snopoke) [inline docker/bootstrap.sh]: Should these still be commented out?

File: docker/bootstrap.sh [hunk 3]
```diff
@@ -2,9 +2,27 @@
...
+if [[ ! $email =~ '\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}\b' ]] ; then
+  echo "Bad email. Try again."
+  exit 1
+fi
+
+echo "Please enter admin password"
+read ADMIN_PASSWORD
+
+echo "OK. Starting services and then bootstrapping. "
+
```
Comment (snopoke) [inline docker/bootstrap.sh]: Nice

File: docker/compose/docker-compose-base.yml [hunk 4]
```diff
@@ -33,6 +33,28 @@ elasticsearch:
...
+#  image: wurstmeister/zookeeper
+#  labels: 
+#    commcare.name: "zookeeper"
+#  expose: 
+#    - "2181"
+
+#kafka:
+#  build: ../dockerfiles/kafka-docker-master
+#  expose:
+#    - "9092"
```
Comment (snopoke) [inline docker/compose/docker-compose-base.yml]: can remove commented lines

File: dev_settings.py (hunk 1 above)
Comment (philipkaare) [inline dev_settings.py]: Right, sorry about that.

File: docker/bootstrap.sh (hunk 2 above)
Comment (philipkaare) [inline docker/bootstrap.sh]: No, you are right. Is there an easy way to fix this from here or should I make a new commit fixing it and the update the PR?

File: docker/compose/docker-compose-base.yml (hunk 4 above)
Comment (philipkaare) [inline docker/compose/docker-compose-base.yml]: I agree.

File: docker/compose/docker-compose-hq.yml [hunk 5]
```diff
@@ -1,11 +1,12 @@
 web:
+  container_name: 'web'
```
Comment (snopoke) [inline docker/compose/docker-compose-hq.yml]: why is this necessary?

File: docker/dockerfiles/Dockerfile_celery [hunk 6]
```diff
@@ -0,0 +1,32 @@
+FROM dimagi/commcarehq_base:latest
+
+RUN apt-get update
+RUN apt-get install -y netcat
```
Comment (snopoke) [inline docker/dockerfiles/Dockerfile_celery]: Is this needed by Celery? Should probably go in the base image.

File: docker/compose/docker-compose-hq.yml (hunk 5 above)
Comment (philipkaare) [inline docker/compose/docker-compose-hq.yml]: I don't think it is after all. I added it for Nginx to be able to link to the container, but since it's also added in dockerhq.sh under runserver-prod, it's redundant here.

File: docker/dockerfiles/Dockerfile_celery [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    && rm -rf /root/.cache/pip
+
+WORKDIR /mnt
+
+RUN mkdir /mnt/shared
+
+ADD docker/entrypoint.sh /mnt/docker/entrypoint.sh
+
+ENTRYPOINT ["/mnt/docker/entrypoint.sh"]
+CMD ["help"]
```
Comment (snopoke) [inline docker/dockerfiles/Dockerfile_celery]: this is almost identical to `Dockerfile_commcarehq`. Is it not worth updating that one rather than building a whole new image?

File: docker/dockerfiles/Dockerfile_celery (hunk 6 above)
Comment (philipkaare) [inline docker/dockerfiles/Dockerfile_celery]: I don't think it is - but I support putting it in the base image. I just added it to make it possible to wait for the web container.

File: docker/dockerfiles/Dockerfile_commcarehq [hunk 8]
```diff
@@ -9,16 +9,39 @@ RUN requirements/uninstall-requirements.sh
...
 
+RUN npm install -g less@1.3.1
+
+WORKDIR /opt
+
+RUN git clone https://github.com/less/less.js.git lessc
+
+WORKDIR /opt/lessc
+
+RUN git reset --hard 546bedd3440ff7e626f629bef40c6cc54e658d7e
```
Comment (snopoke) [inline docker/dockerfiles/Dockerfile_commcarehq]: I think these can be compressed to:
```
RUN git clone https://github.com/less/less.js.git /opt/lessc
  && cd /opt/lessc
  && git reset --hard 546bedd3440ff7e626f629bef40c6cc54e658d7e
```

File: docker/dockerfiles/Dockerfile_commcarehq [hunk 9]
```diff
@@ -9,16 +9,39 @@ RUN requirements/uninstall-requirements.sh
...
+
 WORKDIR /mnt
 
+RUN wget -O jython_installer-2.5.3.jar http://search.maven.org/remotecontent?filepath=org/python/jython-installer/2.5.3/jython-installer-2.5.3.jar 
+RUN ls
+RUN mkdir /usr/local/lib/jython
+RUN java -jar jython_installer-2.5.3.jar -s -d /usr/local/lib/jython
+RUN ln -s /usr/local/lib/jython/bin/jython /usr/local/bin/
+RUN wget http://peak.telecommunity.com/dist/ez_setup.py
+RUN jython ez_setup.py
```
Comment (snopoke) [inline docker/dockerfiles/Dockerfile_commcarehq]: Similar to above, this can be turned into a single multline command.

It should also do some cleanup like removing the installers

File: docker/dockerfiles/Dockerfile_commcarehq [hunk 10]
```diff
@@ -9,16 +9,39 @@ RUN requirements/uninstall-requirements.sh
...
+
+RUN mkdir /mnt/shared
```
Comment (snopoke) [inline docker/dockerfiles/Dockerfile_commcarehq]: Should be `/sharedfiles`

File: dev_settings.py (hunk 1 above)
Comment (snopoke) [inline dev_settings.py]: I see you removed the setting from `dockersettings.py`. Why?

File: docker/localsettings_docker.py [hunk 11]
```diff
@@ -2,13 +2,20 @@
 
 EMAIL_BACKEND = 'django.core.mail.backends.console.EmailBackend'
 
-ADMINS = (('Admin', 'admin@example.com'),)
+ADMINS = (('Admin', 'commcare-admin@bandim.org'),)
```
Comment (snopoke) [inline docker/localsettings_docker.py]: don't need this change

File: docker/localsettings_docker.py [hunk 12]
```diff
@@ -33,4 +40,10 @@
 RESTORE_PAYLOAD_DIR_NAME = 'restore'
 SHARED_TEMP_DIR_NAME = 'temp'
 
-ENABLE_PRELOGIN_SITE = True
+ENABLE_PRELOGIN_SITE = False
+
+KAFKA_URL = 'kafka:9092'
+SHARED_DRIVE_ROOT = '/mnt/sharedfiles'
+
+ENVIRONMENT_HOSTS = {"celery": ["celery"], "all": ["localhost"], "zookeeper": ["kafka"], "postgresql": ["postgresql"], "couchdb": ["couchdb"], "redis": ["redis"], "rabbitmq": ["rabbit"], "kafka": ["kafka"], "ungrouped": [], "webworkers": ["192.168.33.21"], "elasticsearch": ["elasticsearch"], "pillowtop": ["localhost"], "touchforms": ["localhost"], "shared_dir_host": ["localhost"]}
```
Comment (snopoke) [inline docker/localsettings_docker.py]: Shouldn't need this - or at least its not used for anything necessary

File: docker/run_celery.sh [hunk 13]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+
+while ! nc -z rabbit 5672; do echo "Waiting for RabbitMQ... \n"; sleep 3; done
```
Comment (snopoke) [inline docker/run_celery.sh]: Does `entrypoing.sh` not handle this already?

Comment (snopoke) [inline docker/run_celery.sh]: If you add an `DEPENDENT_SERVICES: "RABBIT"` environment variable then I think it should

File: docker/runserver_dev.sh [hunk 14]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
+
+export C_FORCE_ROOT="true"
+export CUSTOMSETTINGS="docker.localsettings_docker"
+
+
+#python manage.py migrate_logs_to_sql
```
Comment (snopoke) [inline docker/runserver_dev.sh]: Comment

File: dockerhq.sh [hunk 15]
```diff
@@ -79,8 +106,31 @@ case $key in
     migrate)
         web_runner run --rm web python manage.py migrate $@
         ;;
-    runserver)
-        web_runner run --service-ports web $@
+    runserver-dev)
+	web_runner run --name commcarehq_web_1 --rm --service-ports web /mnt/docker/runserver_dev.sh
```
Comment (snopoke) [inline dockerhq.sh]: the name should be set automatically if you remove the `custom_name` setting from the compose file

File: settings.py [hunk 16]
```diff
@@ -529,7 +529,7 @@
 GET_URL_BASE = 'dimagi.utils.web.get_url_base'
 
 # celery
-BROKER_URL = 'django://'  # default django db based
+BROKER_URL = 'amqp://guest:guest@rabbit:5672/commcarehq'  # default django db based
```
Comment (snopoke) [inline settings.py]: Doesn't need to be here - overridden in localsettings

File: settings.py [hunk 17]
```diff
@@ -1415,6 +1414,7 @@
     ]
 }
 
+KAFKA_URL = "kafka:9092"
```
Comment (snopoke) [inline settings.py]: also shouldn't be here

File: docker/bootstrap.sh (hunk 2 above)
Comment (snopoke) [inline docker/bootstrap.sh]: Yea, just push a new commit

**Comment by snopoke** [comment]
@philipkaare some more general comments below:

There are currently 3 use cases that we're using docker for:
* travis
* running services locally (to avoid having to install them)
* running the dev server locally for testing

This PR adds 2 more:
* running complete setup locally for testing (with celery / touchforms etc)

This is a pretty uncommon requirement so although its good to make it possible it shouldn't be the default case (which is just running the devserver)

* running prod setup

Although this may be working I don't think it is sufficient for an actual prod setup. It may be better to keep this separate from the HQ repo.

File: docker/run_celery.sh (hunk 13 above)
Comment (philipkaare) [inline docker/run_celery.sh]: I added this because the standard behaviour was waiting on port 5673 (I think it was) and that did not work.

File: dockerhq.sh (hunk 15 above)
Comment (philipkaare) [inline dockerhq.sh]: I had a problem with every run increasing the number after the name, so that's why I added this.

File: docker/run_celery.sh (hunk 13 above)
Comment (snopoke) [inline docker/run_celery.sh]: odd, it gets the port number from the environment variables in the container

File: dockerhq.sh (hunk 15 above)
Comment (snopoke) [inline dockerhq.sh]: That should only happen if the old container never got removed. Could be related to that compose bug you linked to further down?

**Comment by philipkaare** [comment]
I agree that it may be better to make a separate repo for a proper docker production setup.

File: dockerhq.sh (hunk 15 above)
Comment (philipkaare) [inline dockerhq.sh]: I think that's probably the reason. It was certainly nasty to debug.

File: docker/bootstrap.sh (hunk 3 above)
Comment (esoergel) [inline docker/bootstrap.sh]: :+1:

**Comment by esoergel** [comment]
This looks cool, I know I'd use the other services through docker if I could.  Is this PR still active?

**Comment by philipkaare** [comment]
Hi @esoergel I think that @snopoke and I agreed that I would make a new PR which is cleaned up better. I am currently setting it up to run in production in Guinea-Bissau for Bandim Health Project, so once I am done I'll submit my changes and then we can see if anybody can use them.

**Comment by dimagimon** [comment]
There are 9 errors:

* docker/localsettings_docker.py, line 12 - E261 at least two spaces before inline comment
* docker/localsettings_docker_prod.py, line 3 - F401 'sys' imported but unused
* settings.py, line 1621 - W391 blank line at end of file
* docker/localsettings_docker_prod.py, line 12 - E261 at least two spaces before inline comment
* docker/localsettings_docker.py, line 48 - E501 line too long (383 > 115 characters)
* docker/localsettings_docker.py, line 49 - W391 blank line at end of file
* docker/localsettings_docker_prod.py, line 7 - E225 missing whitespace around operator
* docker/localsettings_docker_prod.py, line 67 - W391 blank line at end of file
* docker/localsettings_docker_prod.py, line 62 - E261 at least two spaces before inline comment
E262 inline comment should start with '# '
